lib/rules/prefer-destructuring.js

```diff
@@ -109,24 +136,71 @@ module.exports = {
                 return;
             }
 
-            if (checkArrays && isArrayIndexAccess(rightNode)) {
-                report(reportNode, "array");
+            const sourceCode = context.getSourceCode();
+            const previousNode = sourceCode.getTokenBefore(reportNode);
+
+            // Check to avoid issues with ASI
+            if (previousNode && previousNode.type !== "Keyword" && previousNode.value !== ";") {
```

Comment:
I assume this is intended to check cases like

```js
foo
baz = bar.baz

// fixed to

foo
({ baz } = bar); // <-- ASI issue
```

* The variable name `previousNode` is a bit confusing, because it looks like this is referring to a token, not a node. Could you rename it to something like `previousToken`?
* I'm not sure what `previousNode.type !== "Keyword"`  is used for in this check. It might be better to use [`astUtils.isSemicolonToken`](https://github.com/eslint/eslint/blob/bf4d8cf2ba222bdea5e2ffbe036663cdaf11d278/lib/ast-utils.js#L291-L293) instead.
* Can you add tests for this case?